WalletWasabi/BitcoinCore/Mempool/MempoolMirror.cs

```diff
@@ -0,0 +1,150 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using NBitcoin.Protocol;
+using NBitcoin.RPC;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Bases;
+using WalletWasabi.BitcoinCore.Rpc;
+using WalletWasabi.Blockchain.Transactions;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.BitcoinCore.Mempool
+{
+	public class MempoolMirror : PeriodicRunner
+	{
+		/// <param name="period">How often to mirror the mempool.</param>
+		public MempoolMirror(TimeSpan period, IRPCClient rpc, P2pNode node) : base(period)
+		{
+			Rpc = rpc;
+			Node = node;
+		}
+
+		public IRPCClient Rpc { get; }
+		public P2pNode Node { get; }
+		private Dictionary<uint256, Transaction> Mempool { get; } = new();
+		private object MempoolLock { get; } = new();
+
+		public override Task StartAsync(CancellationToken cancellationToken)
+		{
+			Node.MempoolService.TransactionReceived += MempoolService_TransactionReceived;
+			return base.StartAsync(cancellationToken);
+		}
+
+		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+		{
+			var sw = Stopwatch.StartNew();
+			int added = await MirrorMempoolAsync(stoppingToken).ConfigureAwait(false);
+			Logger.LogInfo($"{added} transactions were copied from the full node to the in-memory mempool within {sw.Elapsed.TotalSeconds} seconds.");
+
+			await base.ExecuteAsync(stoppingToken).ConfigureAwait(false);
+		}
+
+		protected override async Task ActionAsync(CancellationToken cancel)
+		{
+			await MirrorMempoolAsync(cancel).ConfigureAwait(false);
+		}
+
+		public override Task StopAsync(CancellationToken cancellationToken)
+		{
+			Node.MempoolService.TransactionReceived -= MempoolService_TransactionReceived;
+			return base.StopAsync(cancellationToken);
+		}
+
+		private void MempoolService_TransactionReceived(object? sender, SmartTransaction stx)
+		{
+			AddTransactions(stx.Transaction);
+		}
+
+		private int AddTransactions(params Transaction[] txs) => AddTransactions(txs as IEnumerable<Transaction>);
+
+		private int AddTransactions(IEnumerable<Transaction> txs)
+		{
+			var added = 0;
+			lock (MempoolLock)
+			{
+				foreach (var tx in txs.Where(x => !Mempool.ContainsKey(x.GetHash())))
+				{
+					// Evict double spents.
+					EvictSpendersNoLock(tx.Inputs.Select(x => x.PrevOut));
+
+					Mempool.Add(tx.GetHash(), tx);
+					added++;
+				}
+			}
+			return added;
+		}
+
+		private void EvictSpendersNoLock(IEnumerable<OutPoint> txOuts)
```

Comment:
Again, not sure if perf is important here, but this method does really a lot of repetitive work. 
Maybe we can have `HashSet<OutPoint> MempoolTxOuts` and just do something like:

```csharp
HashSet<uint256> doubleSpents = new();
foreach (var input in txOuts)
{
	if (MempoolTxOuts.Contains(input))
	{
		doubleSpents.Add(mempoolTx.Key);
	}	
}
```

to avoid having an algorithm that is linear with the size of the mempool.